packages/flutter_tools/lib/src/run_hot.dart

```diff
@@ -988,3 +928,119 @@ class HotRunner extends ResidentRunner {
     await stopEchoingDeviceLog();
   }
 }
+
+class ProjectFileInvalidator {
+  ProjectFileInvalidator(this._packagesPath, this._flutterProject) {
+    final File packagesFile = fs.file(_packagesPath);
+    if (packagesFile.existsSync()) {
+      _packagesUpdateTime = packagesFile.statSync().modified.millisecondsSinceEpoch;
+      _packageMap = PackageMap(_packagesPath).map;
+    } else {
+      _packagesUpdateTime = -1;
+      _packageMap = const <String, Uri>{};
+    }
+    _computePackageMap(_packageMap, _flutterProject);
+  }
+
+  // Used to avoid watching pubspec directories. This will not change even with pub upgrade,
+  // because that actually switches the directory and requires a corresponding
+  // update to .packages
+  static const String _pubCachePathLinuxAndWindows = '.pub-cache';
+  static const String _pubCachePathWindows = 'Pub/Cache';
+
+  Map<String, Uri> _packageMap;
+  final String _packagesPath;
+  final FlutterProject _flutterProject;
+  final Map<String, int> _updateTime = <String, int>{};
+  int _packagesUpdateTime;
+
+  @visibleForTesting
+  Map<String, int> get updateTime => _updateTime;
+
+  @visibleForTesting
+  Map<String, Uri> get packageMap => _packageMap;
+
+  static void _computePackageMap(Map<String, Uri> packageMap, FlutterProject flutterProject) {
+    if (flutterProject != null && flutterProject.pubspecFile.existsSync()) {
+      try {
+        final PubspecYaml pubspec = PubspecYaml(flutterProject.directory);
+        final Set<String> relevantDependencies = <String>{};
+        for (PubspecDependency dependency in pubspec.allDependencies) {
+          if (dependency.isDevDependency) {
+            continue;
+          }
+          relevantDependencies.add(dependency.name);
+        }
+        // Remove any packages which were tagged as dev dependenices,
+        // But don't remove the app itself!
+        for (String packageName in packageMap.keys.toList()) {
+          if (!relevantDependencies.contains(packageName) && packageName != flutterProject.manifest.appName) {
+            packageMap.remove(packageName);
+            continue;
+          }
+        }
+      } catch (err) {
+        // If we detect a pubspec formatting problem, fallback to the packages file.
+      }
+    }
+    // Remove any packages which are derived from the pub cache.
+    for (String packageName in packageMap.keys.toList()) {
+      final String path = packageMap[packageName].path;
+      if ((platform.isWindows && path.contains(_pubCachePathWindows))
+          || path.contains(_pubCachePathLinuxAndWindows)) {
+        packageMap.remove(packageName);
+      }
+    }
+  }
+
+  // To initialize the file invalidator we traverse all of the source files
+  // to grab an initial timestamp. We assume that the initial dill file
+  // contains all up to date files.
```

Comment:
I'm not sure where we use this assumption, but generally speaking I don't think it's a safe to assume this.
User might have updated some sources after dill file was produced and packaged for execution and `flutter run` with incremental compiler got going.

Reply:
Looking at the existing hot reload code, we do assume that all files are initially dirty. I'll update this code to match that assumption